### PR TITLE
[zero] ZeRO supports pipeline parallel

### PR DESCRIPTION
**Pull request by ver217**

### CODE REVIEW

File: colossalai/engine/schedule/_pipeline_schedule.py [hunk 1]
```diff
@@ -79,7 +80,7 @@ def load_batch(self, data_iter):
 
     def _get_data_slice(self, data, offset):
         if isinstance(data, torch.Tensor):
-            return data[offset: offset + self.microbatch_size]
+            return data[offset:offset + self.microbatch_size]
         else:
```
Comment (feifeibear) [inline colossalai/engine/schedule/_pipeline_schedule.py:84]: elif isinstance(data, dict):

File: colossalai/engine/schedule/_pipeline_schedule.py [hunk 2]
```diff
@@ -92,13 +93,14 @@ def load_micro_batch(self):
...
         if isinstance(engine.optimizer, ShardedOptimizer) or isinstance(engine.model, ShardedModel):
-            raise TypeError(
-                "Pipeline schedule is currently not compatible with ZeRO"
-            )
+            raise TypeError("Pipeline schedule is currently not compatible with ZeRO")
         model = engine.model
         if isinstance(model, NaiveAMPModel):
             self.dtype = torch.half
             model = model.model
+        elif isinstance(model, ShardedModelV2):
```
Comment (feifeibear) [inline colossalai/engine/schedule/_pipeline_schedule.py:101]: why not merge it with the above if?

File: colossalai/engine/schedule/_pipeline_schedule.py [hunk 3]
```diff
@@ -107,6 +109,8 @@ def pre_processing(self, engine):
     def _call_engine(model, input_tensor, batch_data):
         if isinstance(model, NaiveAMPModel):
             sig = inspect.signature(model.model.forward)
+        elif isinstance(model, ShardedModelV2):
```
Comment (feifeibear) [inline colossalai/engine/schedule/_pipeline_schedule.py:112]: why not merge with above if

File: colossalai/engine/schedule/_pipeline_schedule.py (hunk 1 above)
Comment (ver217) [reply colossalai/engine/schedule/_pipeline_schedule.py]: done

File: colossalai/engine/schedule/_pipeline_schedule.py (hunk 2 above)
Comment (ver217) [reply colossalai/engine/schedule/_pipeline_schedule.py]: done

File: colossalai/engine/schedule/_pipeline_schedule.py (hunk 3 above)
Comment (ver217) [reply colossalai/engine/schedule/_pipeline_schedule.py:109]: These two branches are different.

**Review by feifeibear — APPROVED** [review]

**Pull request merged**